query/src/client/QueryMetadataEditor/QueryMetadataEditor.tsx

```diff
@@ -0,0 +1,374 @@
+            domain: newDomain,
+            dirty: state.dirty || dirty, // if the state is already dirty, leave it as such
+            showSave: true
+        }));
+    };
+
+    showMessage = (message: string, messageType: string, index: number, additionalState?: Partial<IAppState>) => {
+        const { messages } = this.state;
+
+        this.setState(Object.assign({}, additionalState, {
+            messages : messages.set(index, {message: message, messageType: messageType})
+        }));
+    };
+
+    dismissChangeConfirmation = () => {
+        const {showViewDataConfirmationModal, showEditSourceConfirmationModal} = this.state;
+
+        this.setState(() => ({
+            showResetConfirmationModal: false,
+            showViewDataConfirmationModal: false,
+            showEditSourceConfirmationModal: false,
+            dirty: false
+        }));
+
+        if (showViewDataConfirmationModal) {
+            this.navigateToViewData();
+        }
+
+        if (showEditSourceConfirmationModal) {
+            this.navigateToEditSource();
+        }
+    };
+
+    dismissAlert = (index: any) => {
+        this.setState(() => ({
+            messages: this.state.messages.setIn([index], [{message: undefined, messageType: undefined}])
+        }));
+    };
+
+    aliasFieldBtnHandler = () => {
+        this.setState(() => ({showAlias: true}));
+    };
+
+    onHideAliasField = (): any => {
+        this.setState(() => ({showAlias: false}));
+    };
+
+    onAddAliasField = (domainField: DomainField): any => {
+        const { domain } = this.state;
+        const newFields = domain.fields.push(domainField);
+        const newDomain = domain.set('fields', newFields) as DomainDesign;
+
+        this.setState(() => ({
+            showAlias: false,
+            domain: newDomain,
+            showSave: true,
+            dirty: true
+        }));
+    };
+
+    onConfirmEditSource = () => {
+        const { dirty } = this.state;
+
+        if (dirty) {
+            this.onSaveBtnHandler();
+        }
+
+        this.navigateToEditSource();
+    };
+
+    navigateToViewData() {
+        const { schemaName, queryName } = this.state;
+
+        this.setState(() => ({dirty: false}), () => {
+            window.location.href =  ActionURL.buildURL('query', 'executeQuery', LABKEY.container.path, {schemaName: schemaName, ['query.queryName']: queryName});
+        });
+    }
+
+    onConfirmViewData = () => {
+        const { dirty } = this.state;
+
+        if (dirty) {
+            this.onSaveBtnHandler();
+        }
+
+        this.navigateToViewData();
+    };
+
+    onSaveBtnHandler = () => {
+        const { domain, schemaName, messages } = this.state;
+
+        saveQueryMetadata(domain, schemaName)
+            .then(() => {
+                this.showMessage("Save Successful", 'success', 0);
+                this.setState(() => ({
+                    showSave: false,
+                    dirty: false
+                }));
+            })
```

Comment:
If you did the callback suggestion from above, I think this function would look something like this.

```suggestion
    onSaveBtnHandler = (onSaveNavigation) => {
        const { domain, schemaName, messages } = this.state;

        saveQueryMetadata(domain, schemaName)
            .then(() => {
                this.showMessage("Save Successful", 'success', 0);
                this.setState(() => ({
                    showSave: false,
                    dirty: false
                }), () => {
                    onSaveNavigation();
                });
            })
```